ramips: add support for TP-Link RE200v1

andyboeh opened the pull request:
TP-Link RE200v1 is a wireless range extender with Ethernet and 2.4G and 5G
WiFi with internal antennas. It's based on MediaTek MT7620A+MT7610EN.

### Specification:

- MediaTek MT7620A (580 Mhz)
- 64 MB of RAM
- 8 MB of FLASH
- 2T2R 2.4 GHz and 1T1R 5 GHz
- 1x 10/100 Mbps Ethernet
- UART header on PCB (57600 8n1)
- 8x LED (GPIO-controlled; only 6 supported), 2x button

There are 2.4G and 5G LEDs in red and green which are controlled
separately. The 5G LED is currently not supported, since the GPIOs couldn't
be determined.

### Installation:

#### Web Interface
It is possible to upgrade to OpenWrt via the web interface. However, the
OEM firmware upgrade file is required and a tool to fix the MD5 sum of
the header. This procedure overwrites U-Boot and there is no failsafe / 
recovery mode present! To prepare an image, you need to take the header
and U-Boot (i.e. 0x200 + 0x20000 bytes) from an OEM firmware file and
attach the factory image to it. Then fix the header MD5Sum1.

#### Serial console
Opening the case is quite hard, since it is welded together. Rename the
OpenWrt factory image to "test.bin", then plug in the device and quickly
press "2" to enter flash mode (no line feed). Follow the prompts until
OpenWrt is installed.

Unfortunately, this devices does not offer a recovery mode or a tftp
installation method. If the web interface upgrade fails, you have to open
your device and attach serial console. Since the web upgrade overwrites
the boot loader, you might also brick your device.

### Additional notes:

This seems to be the first ramips device with a TP-Link v1 header. The
original firmware has the string "EU" embedded, there might be some region-
checking going on during the firmware upgrade process. The original
firmware also contains U-Boot and thus overwrites the boot loader during
upgrade.
In order to flash back to stock, the first header and U-Boot needs to be
stripped from the original firmware.

andyboeh committed:
ramips: add support for TP-Link RE200 v1

TP-Link RE200 v1 is a wireless range extender with Ethernet and 2.4G and 5G
WiFi with internal antennas. It's based on MediaTek MT7620A+MT7610EN.

Specifications
--------------

- MediaTek MT7620A (580 Mhz)
- 64 MB of RAM
- 8 MB of FLASH
- 2T2R 2.4 GHz and 1T1R 5 GHz
- 1x 10/100 Mbps Ethernet
- UART header on PCB (57600 8n1)
- 8x LED (GPIO-controlled; only 6 supported), 2x button

There are 2.4G and 5G LEDs in red and green which are controlled
separately. The 5G LED is currently not supported, since the GPIOs couldn't
be determined.

Installation
------------

Web Interface
-------------

It is possible to upgrade to OpenWrt via the web interface. However, the
OEM firmware upgrade file is required and a tool to fix the MD5 sum of
the header. This procedure overwrites U-Boot and there is not failsafe /
recovery mode present! To prepare an image, you need to take the header
and U-Boot (i.e. 0x200 + 0x20000 bytes) from an OEM firmware file and
attach the factory image to it. Then fix the header MD5Sum1.

Serial console
--------------

Opening the case is quite hard, since it is welded together. Rename the
OpenWrt factory image to "test.bin", then plug in the device and quickly
press "2" to enter flash mode (no line feed). Follow the prompts until
OpenWrt is installed.

Unfortunately, this devices does not offer a recovery mode or a tftp
installation method. If the web interface upgrade fails, you have to open
your device and attach serial console. Since the web upgrade overwrites
the boot loader, you might also brick your device.

Additional notes
----------------

MAC address assignment is based on stock-firmware. For me, the device
assigns the MAC on the label to Ethernet and the 2.4G WiFi, while the 5G
WiFi has a separate MAC with +2.

*:88    Ethernet/2.4G    label, uboot 0x1fc00, userconfig 0x0158
*:89    unused           userconfig 0x0160
*:8A    5G               not present in flash

This seems to be the first ramips device with a TP-Link v1 header. The
original firmware has the string "EU" embedded, there might be some region-
checking going on during the firmware upgrade process. The original
firmware also contains U-Boot and thus overwrites the boot loader during
upgrade.
In order to flash back to stock, the first header and U-Boot need to be
stripped from the original firmware.

Signed-off-by: Andreas Böhler <dev@aboehler.at>

adrianschmutzler commented:
Hi, your PR has several formal issues. Please have another look at https://openwrt.org/submitting-patches.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,158 @@
+/dts-v1/;
```
Please add a license.

andyboeh replied:
Should be fixed.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,158 @@
+		
+		led_power: power {
+		    label = "re200-v1:green:power";
+		    gpios = <&gpio1 0 GPIO_ACTIVE_LOW>;
```
Please make your indent consistent in the entire file.

andyboeh replied:
Yeah, I knew this would happen. I usually use spaces for all other projects...

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,158 @@
+		    gpios = <&gpio1 15 GPIO_ACTIVE_LOW>;
+		};
+		
+		led_2g1 {
```
What's this LED for?

andyboeh replied:
Green 2.4G LED, the other one is red. Original firmware shows the WiFi signal strength here.

adrianschmutzler replied:
The use "wlan2_green" and "wlan2_red" or "wifi2_green" and "wifi2_red" or "wlan2g_green" etc.
Based on your choice of the name, also use "wifi" or "wlan" for the separate LED.

andyboeh replied:
done

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,158 @@
+			gpios = <&gpio2 0 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan {
```
2g, 5g, or both?

andyboeh replied:
Neither, it's a separate LED (there are 5 WiFi LEDs, one general for WiFi, i.e. this one, green/red 2.4G and green/red 5G).

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,158 @@
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
```
flash@0

andyboeh replied:
I copied that from another device, fixed.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,158 @@
+	m25p80@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Please check whether this can be raised.

andyboeh replied:
How do I do that? Trial and error?

adrianschmutzler replied:
You try to find some datasheet stating it, have a look at what similar devices do, or ask people that might know it.

achterin replied:
Found a datasheet that states the following:
- 104MHz clock rate for one data bit
- 80MHz clock rate for two data bits
- 50MHz clock rate for four data bits

andyboeh replied:
@achterin I found a datasheet, too, that states 50 MHz for two data bits, the rest is the same.
I suppose raising it to 50 MHz is safe?

achterin replied:
@andyboeh yeah, i think 50MHz should be safe to use here

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,158 @@
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&radio 0>;
```
Is there a MAC address in radio 0x4 and/or 0x8004?

andyboeh replied:
Hard to tell, definitely not the one written at the backside. The MACs I'm using are more or less consistent with original firmware (they assign the same MAC to one of the radios and the LAN device, but they skip the MAC between the two).

adrianschmutzler replied:
Have a look at: https://openwrt.org/docs/guide-developer/mac.address

andyboeh replied:
I searched the whole flash for the MAC address printed at the bottom of the device and I could only find it in the userconfig partition at the locations present in the dts. 
There are three addresses:
*:88
*:89
*:89

When running stock firmware, my device assigns *:88 to LAN and 2.4G WiFi and *:8A to the 5G WiFi. Thus, I use *:88 for LAN, *:89 for 2.4G and *:8A for 5G WiFi.

adrianschmutzler replied:
So, you have the following?:
```
MAC assignment based on stock firmware:
*:88    Ethernet/2.4G    userconfig 0x0158
*:89    unused         userconfig 0x0160
*:8A    5G                 not present in flash
```

If it is correct, please add this to your commit message (and maybe format it more nicely.

Please stick to the MAC address assignment of the stock firmware. You can never be sure whether the additional unused MAC address will be there on other devices. For the same reason, I would calculated 5G Wifi based on 0x158, so adding +2 instead of +1.

Normally, userconfig partition is for stuff created by OpenWrt firmware. So, I would expect to find one of the MAC address in u-boot, too, and would base everything on that one (if it really exists).

Which of the addresses is on the label?

andyboeh replied:
Correct, the *:88 is on the label. 
Yes, the *:88 is also in the U-Boot partition at 0x1fc00 - seems to be a safe guess then. I changed it as you suggested.

adrianschmutzler replied:
Perfect. That would be the standard TP-Link position I expected.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -869,6 +892,29 @@ define Device/tplink_archer-c20i
 
+define Device/tplink_re200-v1
+  MTK_SOC := mt7620a
+  DEVICE_VENDOR := TP-Link
```
Please keep DEVICE_ variables together.

andyboeh replied:
done

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -869,6 +892,29 @@ define Device/tplink_archer-c20i
+  DEVICE_VENDOR := TP-Link
+  TPLINK_BOARD_ID := RE200-V1
+  TPLINK_HWID := 0x02000001
+  TPLINK_HWREV := 0x00000001
```
0x1?

andyboeh replied:
Should be fixed.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -869,6 +892,29 @@ define Device/tplink_archer-c20i
+  KERNEL_ENTRY := 0x8000c310
+  KERNEL_LOADADDR := 0x80000000
+  DEVICE_VARIANT := v1
+  SUPPORTED_DEVICES += re200
```
Drop this line.

andyboeh replied:
You mean the `SUPPORTED_DEVICES +=` line? Tbh, I have no clue what this does, I just copied it from the other devices.

adrianschmutzler replied:
I know ...

adrianschmutzler commented on target/linux/ramips/mt7620/base-files/etc/board.d/01_leds:
```diff
@@ -198,6 +198,10 @@ tplink,archer-mr200)
 	;;
+tplink,re200-v1)
+	ucidef_set_led_netdev "lan" "lan" "$boardname:green:lan" "eth0"
+	set_wifi_led "$boardname:green:wlan"
```
Can you use a DTS trigger instead?

andyboeh replied:
For WiFi I should've fixed it, how to do it for LAN?

adrianschmutzler replied:
LAN has to be kept here.

andyboeh commented:
> Hi, your PR has several formal issues. Please have another look at https://openwrt.org/submitting-patches.

I hope I got everything alright now, if not, could you tell me what I missed, please?

adrianschmutzler commented:
> > Hi, your PR has several formal issues. Please have another look at https://openwrt.org/submitting-patches.
> 
> I hope I got everything alright now, if not, could you tell me what I missed, please?

Actually, you have addressed none of the formal issues at all. Some selected things:
- squash all changes into a single commit
- proper commit title
- proper commit message incl. device specifications and installation
- Signed-off-by

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -869,6 +892,28 @@ define Device/tplink_archer-c20i
 TARGET_DEVICES += tplink_archer-c20i
 
+define Device/tplink_re200-v1
+  MTK_SOC := mt7620a
```
Rename to "SOC" due to 58e71e322bc0e0edd08157075e9022c15ae33ec5.

andyboeh commented:
The 5G LEDs are at register 0xc0d80228. Does anybody know where this register points to or how I can toggle the LEDs there?

andyboeh commented:
> > > Hi, your PR has several formal issues. Please have another look at https://openwrt.org/submitting-patches.
> > 
> > 
> > I hope I got everything alright now, if not, could you tell me what I missed, please?
> 
> Actually, you have addressed none of the formal issues at all. Some selected things:
> 
>     * squash all changes into a single commit
> 
>     * proper commit title
> 
>     * proper commit message incl. device specifications and installation
> 
>     * Signed-off-by

Thanks, I'll fix that once the other details are fixed.

andyboeh commented:
I *think* I fixed the remaining issues, apart from squashing of the commits etc. I changed the SPI frequency to 50MHz, this seems to work fine.
Still missing are the 5G LEDs at 0xc0d80228 where I have no clue what register this is.

achterin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -878,6 +901,28 @@ define Device/tplink_archer-c20i
+  IMAGE/factory.bin := mktplinkfw factory -e -O
+  IMAGE_SIZE := 7936k
+  KERNEL_ENTRY := 0x8000c310
+  KERNEL_LOADADDR := 0x80000000
```
According to the vendor fw these values are as follows:
Kernel load address: 0x00000080
Kernel entry point: 0x10c30080
did i miss something or are these just copied from another device?

andyboeh replied:
During boot of original firmware, exactly those values are reported on Serial and they are the same as mktplinkfw -i reports on the OEM firmware. However, the endianness needs to be converted, this is achieved with the "-e" switch during firmware creation.

achterin replied:
ah ok, thanks for the hint. i am currently working on the RE210v1, which is quite similar. i need to calculate my values again then :+1:

andyboeh replied:
ah, sorry, you're right, it displays your values, not mine - at least the same for the generated firmware and the OEM firmware.
I suppose it would work, but why do I want to pass wrong values when I can pass the correct ones and have the tool swap endianness?
This is what U-Boot reports on serial:
```
## Booting image at bc020000 ...
   Uncompressing Kernel Image ... OK
No initrd
## Transferring control to Linux (at address 8000c310) ...
## Giving linux memsize in MB, 64

Starting kernel ...
```

achterin replied:
yeah, there is no point in that. but i have another idea that might shrink your code down a bit. in mktplinkfw.c you can find a flash_layout struct. you might want to add another entry there containing your values as my device is using the same kernel addresses. that way, you can shorten the mktplinkfw define and just have to adjust the TPLINK_FLASHLAYOUT variable as "8Mlzma" is clearly not correct for both our devices. if you dont wannt the deal with that, i can adjust it in my pull request later on which would only require retesting your device.

andyboeh replied:
Good point! I thought about that previously, but I didn't want to modify the tool for just one device.

Somebody is working ok the RE200v3 which might also use the same layout. 

Now it makes sense, so I'll adapt the tool later today.

andyboeh replied:
I added the new layout "8Mmtk", because this name is also used in mktplinkfw2.c for MediaTek chipsets. It works fine for me :)

achterin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -878,6 +901,28 @@ define Device/tplink_archer-c20i
+  KERNEL_LOADADDR := 0x80000000
+  KERNEL := $(KERNEL_DTB)
+  KERNEL_INITRAMFS := $(KERNEL_DTB) | tplink-v1-header -e -O
+  TPLINK_BOARD_ID := RE200-V1
```
this is not needed as the device doesn't use tplink-safeloader

andyboeh replied:
There are four lines displayed, what exactly do you mean?

adrianschmutzler replied:
the one directly above.

andyboeh replied:
OK, done.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,160 @@
+		wlan {
+			label = "re200-v1:green:wlan";
+			gpios = <&gpio1 3 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0radio";
```
If this is the general LED, I would not assign any trigger to it, but assign the appropriate triggers for 2.4 GHz and 5 GHz to the green 2g and 5g LEDs.

andyboeh replied:
Yeah, problem is: I couldn't determine the 5G GPIOs. I *assume* they are connected to the mt7610e, but I couldn't verify it.

adrianschmutzler replied:
Still, the trigger for 2.4 GHz should be assigned to 2.4 GHz LED, no matter whether it is phy0radio or phy0tpt after all.

andyboeh replied:
I changed it to phy1radio, because phy0 is actually the 5G radio (there is no phy1tpt trigger).

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,160 @@
+
+		wlan2g_green {
+			label = "re200-v1:green:wlan2g";
+			gpios = <&gpio3 0 GPIO_ACTIVE_LOW>;
```
This is where I would put the trigger for 2g Wifi. Did you check whether phy0tpt works, as this would be preferred over phy0radio, which shows no activity (blinking).

andyboeh replied:
No, I didn't check it. IMHO phy0tpt makes sense for the general LED (as mentioned above) to indicate WiFi active or not and the radio triggers would make sense on the green 2G/5G LEDs - but the same problem with the 5G WiFi applies.

adrianschmutzler replied:
Still, if there is a 2g led, I would use that with the appropriate trigger (phy0tpt or phy1tpt) if it is available on your device.

BTW, which is which, 2G and 5G vs. phy0 and phy1?

Regarding 5G, have you checked this approach:
https://github.com/openwrt/openwrt/blob/38bdfcdb87e9b3a1232e9820bd6534da4c28282b/target/linux/ramips/dts/mt7620a_iptime.dtsi#L74-L77 ?

Maybe @mans0n can help here, too.

mans0n replied:
Try `led-sources = <2>;`. If the LED turns on even when wlan is off, then add `led-active-low;`
I haven't seen a device with two led per phy, so I'm not sure how to add both red/green leds.

andyboeh replied:
Didn't work, as mentioned below, I gave up on this.

adrianschmutzler replied:
So, you really tried this:

```
&pcie0 {
	mt76@0,0 {
		reg = <0x0000 0 0 0 0>;
		mediatek,mtd-eeprom = <&radio 0x8000>;
		mtd-mac-address = <&uboot 0x1fc00>;
		mtd-mac-address-increment = <2>;
		ieee80211-freq-limit = <5000000 6000000>;

		led {
			led-sources = <2>;
			led-active-low;
		};
	};
};
```

?

andyboeh replied:
Yes, didn't work.
I flashed back to stock: the 5G LEDs don't work either. Maybe my device is bad?

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,160 @@
+};
+
+&gpio0 {
+    status = "okay";
```
Indent.

adrianschmutzler replied:
gpio0 is enabled by default, so you can _remove_ the whole node here.

andyboeh replied:
Done

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,160 @@
+
+&state_default {
+	gpio {
+		ralink,group = "i2c", "uartf", "ephy", "wled", "nd_sd", "rgmii1", "spi refclk";
```
I do not find the reason for nd_sd?

andyboeh replied:
Probably a left-over from trying to identify the LED pins.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,160 @@
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&radio 0x8000>;
+		mtd-mac-address = <&userconfig 0x0160>;
+		mtd-mac-address-increment = <1>;
```
See above.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -16,6 +16,27 @@ define Build/elecom-header
+# -X reserve <size> bytes in the firmware image (hexval prefixed with 0x)
+define Build/mktplinkfw
+	-$(STAGING_DIR_HOST)/bin/mktplinkfw \
+		-H $(TPLINK_HWID) -W $(TPLINK_HWREV) -F $(TPLINK_FLASHLAYOUT) -N OpenWrt -V $(REVISION) \
```
That's more than 80 characters. I would do line break more evenly:
```
define Build/mktplinkfw
	-$(STAGING_DIR_HOST)/bin/mktplinkfw \
		-H $(TPLINK_HWID) -W $(TPLINK_HWREV) -F $(TPLINK_FLASHLAYOUT) \
		-N OpenWrt -V $(REVISION) -m $(TPLINK_HEADER_VERSION) \
		-k $(IMAGE_KERNEL) -r $(IMAGE_ROOTFS) -o $@.new -j -X 0x40000 -a 0x4 \
		$(wordlist 2,$(words $(1)),$(1)) \
		$(if $(findstring sysupgrade,$(word 1,$(1))),-s) && mv $@.new $@ || rm -f $@
endef
```

andyboeh replied:
Done.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -878,6 +899,25 @@ define Device/tplink_archer-c20i
+define Device/tplink_re200-v1
+  SOC := mt7620a
+  DEVICE_VENDOR := TP-Link
+  DEVICE_PACKAGES := kmod-mt76x0e
```
DEVICE_VENDOR, _MODEL, _VARIANT should be together. DEVICE_PACKAGES may be moved somewhere after them.

andyboeh replied:
Fixed.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -878,6 +899,25 @@ define Device/tplink_archer-c20i
+  DEVICE_PACKAGES := kmod-mt76x0e
+  DEVICE_MODEL := RE200
+  DEVICE_VARIANT := v1
+  IMAGES+= factory.bin
```
Add space before +=

andyboeh replied:
Done.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,156 @@
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&radio 0>;
```
Please use <&radio 0x0> here for consistency.

adrianschmutzler replied:
So, please add the following to your commit message:
```
MAC assignment based on stock firmware:
*:88    Ethernet/2.4G    label, uboot 0x1fc00, userconfig 0x0158
*:89    unused         userconfig 0x0160
*:8A    5G                 not present in flash
```

andyboeh replied:
Done

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -0,0 +1,156 @@
+		led-boot = &led_power;
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
```
You can add `label-mac-device = &ethernet;` here.

andyboeh replied:
Done

adrianschmutzler commented:
Nitpick: Add a space between RE200 and v1 in your commit title.

mans0n commented:
FYI wlan2g_green will work with `pinctrl-0 = <&wled_pins>;`.(See mt7620a_iptime.dtsi) This does not need any trigger, the hardware will take care of it. But that might also take away some control on the LED from user, so the choice is yours.

mans0n commented:
I guess the lan led turns on even without the cable, doesn't it? I think the led will work automatically even if you remove the lan led definition and "ephy" on the pinctrl gpio list...

andyboeh commented:
> I guess the lan led turns on even without the cable, doesn't it? I think the led will work automatically even if you remove the lan led definition and "ephy" on the pinctrl gpio list...

No, the LAN led works as intended.
It does work with removing ephy, but it flashes along with the power LED during bootup.

andyboeh commented:
I'm giving up on the 5G LEDs. They're not defined in U-Boot, not defined in the GPL kernel source and I can't find them.

achterin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -878,6 +895,25 @@ define Device/tplink_archer-c20i
 endef
 TARGET_DEVICES += tplink_archer-c20i
 
+define Device/tplink_re200-v1
```
please move the definition to the proper alphabetical position (behind Device/tplink_archer-mr200)

andyboeh replied:
Done

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -947,6 +964,25 @@ define Device/vonets_var11n-300
 endef
 TARGET_DEVICES += vonets_var11n-300
 
+define Device/tplink_re200-v1
```
This should be before vonets_var11n-300, shouldn't it?

andyboeh replied:
Whops, yeah, of course!

adrianschmutzler commented:
I merged your patch, thanks.

There are some minor things where you/we could work on:
- Recently, support for phyXtpt triggers was added to master. It would be nice if you could take latest master, just change phy1radio to phy1tpt and try whether you get activity LEDs then.
- With this one merged, we can now try to create a common DTSI in the PR for RE210
- I will attempt to create tplink-v1-image in image-commands.mk, as tplink-v1-header of header/image for v2 are already there.

adrianschmutzler commented:
@andyboeh Can you please test my patch "move mktplinkfw to tplink-v1-image in image-commands.mk" as found in my staging on your device:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=commit;h=05d3db6dbd9c565aba7163a3b7c52d623d809de5

You might either pick the patch or just build my staging tree for that ...

andyboeh commented:
@adrianschmutzler of course, I just can't promise that it's going to happen before next weekend.

andyboeh commented:
@adrianschmutzler I was about to test your changes when I noticed that they have already been merged. I tested the current snapshot build instead, it works fine via sysupgrade.

adrianschmutzler commented:
Yes, since achterin tested on his RE210, I assumed yours would work, too.

But thanks for confirming that anyway, you can never be sure enough :-)

nos1609 commented:
Not ready for tool-less user install. Should be fixed either with mktplinkfw binary for x86_64 linux, either with a suitable transition firmware for both EU and US regions like for similar tp-link devices on mt7620 either with some updated version of such [tool](https://github.com/lark/fixsum).
Till there is no way to be installed without opening the device this should not be present in snapshots.

andyboeh commented:
There *is* a relatively simple way to install without opening the case and the steps are documented here and in the Wiki. It is very unlikely that there will be a tool-less user install, unless someone creates a custom U-Boot for the device.
In order to ease user installation, I'm in the process of creating a simple GUI tool that performs the necessary firmware modifications.
